File: client/src/cli/ontology.ts
/**
 * Ontology Management Commands
 */

import { Command } from 'commander';
import { createClientFromEnv } from '../api/client';
import * as colors from './colors';
import { coloredCount, separator } from './colors';
import { Table } from '../lib/table';

export const ontologyCommand = new Command('ontology')
  .alias('onto')  // Short alias
  .description('Manage ontologies (knowledge domains)')
  .showHelpAfterError('(add --help for additional information)')
  .showSuggestionAfterError()
  .addCommand(
    new Command('list')
      .description('List all ontologies')
      .action(async () => {
        try {
          const client = createClientFromEnv();
          const result = await client.listOntologies();

          if (result.count === 0) {
            console.log(colors.status.warning('\n⚠ No ontologies found'));
            return;
          }

          console.log('\n' + colors.ui.title('📚 Ontologies in Knowledge Graph'));

          // Use Table system for consistent formatting
          const table = new Table({
            columns: [
              {
                header: 'Ontology',
                field: 'ontology',
                type: 'heading',
                width: 'flex',
                priority: 3
              },
              {
                header: 'Files',
                field: 'file_count',
                type: 'count',
                width: 10,
                align: 'right'
              },
              {
                header: 'Chunks',
                field: 'source_count',
                type: 'count',
                width: 10,
                align: 'right'
              },
              {
                header: 'Concepts',
                field: 'concept_count',
                type: 'count',
                width: 12,
                align: 'right'
              }
            ]
          });

          table.print(result.ontologies);
        } catch (error: any) {
          console.error(colors.status.error('✗ Failed to list ontologies'));
          console.error(colors.status.error(error.response?.data?.detail || error.message));
          process.exit(1);
        }
      })
  )
  .addCommand(
    new Command('info')
      .description('Get detailed information about an ontology')
      .showHelpAfterError()
      .argument('<name>', 'Ontology name')
      .action(async (name) => {
        try {
          const client = createClientFromEnv();
          const info = await client.getOntologyInfo(name);

          console.log('\n' + separator());
          console.log(colors.ui.title(`📖 Ontology: ${name}`));
          console.log(separator());

          console.log('\n' + colors.stats.section('Statistics'));
          console.log(`  ${colors.stats.label('Files:')} ${coloredCount(info.statistics.file_count)}`);
          console.log(`  ${colors.stats.label('Chunks:')} ${coloredCount(info.statistics.source_count)}`);
          console.log(`  ${colors.stats.label('Concepts:')} ${coloredCount(info.statistics.concept_count)}`);
          console.log(`  ${colors.stats.label('Evidence:')} ${coloredCount(info.statistics.instance_count)}`);
          console.log(`  ${colors.stats.label('Relationships:')} ${coloredCount(info.statistics.relationship_count)}`);

          if (info.files.length > 0) {
            console.log('\n' + colors.ui.header('Files'));
            console.log(separator(80, '─'));
            info.files.forEach(file => {
              console.log(`  ${colors.ui.bullet('●')} ${colors.evidence.document(file)}`);
            });
          }
          console.log('\n' + separator());
        } catch (error: any) {
          console.error(colors.status.error('✗ Failed to get ontology info'));
          console.error(colors.status.error(error.response?.data?.detail || error.message));
          process.exit(1);
        }
      })
  )
  .addCommand(
    new Command('files')
      .description('List files in an ontology')
      .showHelpAfterError()
      .argument('<name>', 'Ontology name')
      .action(async (name) => {
        try {
          const client = createClientFromEnv();
          const result = await client.getOntologyFiles(name);

          console.log('\n' + separator());
          console.log(colors.ui.title(`📁 Files in: ${name}`));
          console.log(separator());
          console.log(colors.status.success(`\n✓ Found ${result.count} files:\n`));

          result.files.forEach(file => {
            console.log(colors.evidence.document(file.file_path));
            console.log(`  ${colors.ui.key('Chunks:')} ${coloredCount(file.chunk_count)}`);
            console.log(`  ${colors.ui.key('Concepts:')} ${coloredCount(file.concept_count)}`);
            console.log();
          });
          console.log(separator());
        } catch (error: any) {
          console.error(colors.status.error('✗ Failed to get ontology files'));
          console.error(colors.status.error(error.response?.data?.detail || error.message));
          process.exit(1);
        }
      })
  )
  .addCommand(
    new Command('rename')
      .description('Rename an ontology')
      .showHelpAfterError()
      .argument('<old-name>', 'Current ontology name')
      .argument('<new-name>', 'New ontology name')
      .option('-y, --yes', 'Skip confirmation prompt')
      .action(async (oldName, newName, options) => {
        try {
          if (!options.yes) {
            console.log('\n' + separator());
            console.log(colors.status.warning('⚠️  Rename Ontology'));
            console.log(separator());
            console.log(`\nRename: ${colors.concept.label(oldName)} → ${colors.concept.label(newName)}`);
            console.log(`This will update all Source nodes in the ontology.`);
            console.log('\nUse ' + colors.ui.key('-y') + ' or ' + colors.ui.key('--yes') + ' flag to proceed\n');
            return;
          }

          const client = createClientFromEnv();
          const result = await client.renameOntology(oldName, newName);

          console.log('\n' + separator());
          console.log(colors.status.success(`✓ Renamed ontology "${result.old_name}" → "${result.new_name}"`));
          console.log(separator());
          console.log(`  ${colors.ui.key('Sources updated:')} ${coloredCount(result.sources_updated)}`);
          console.log('\n' + separator());
        } catch (error: any) {
          console.error(colors.status.error('✗ Failed to rename ontology'));
          const detail = error.response?.data?.detail || error.message;
          console.error(colors.status.error(detail));

          // Provide helpful hints based on error
          if (detail.includes('does not exist')) {
            console.error(colors.status.dim('\n  Hint: Use "kg ontology list" to see available ontologies'));
          } else if (detail.includes('already exists')) {
            console.error(colors.status.dim('\n  Hint: Choose a different name or delete the existing ontology first'));
          }

          process.exit(1);
        }
      })
  )
  .addCommand(
    new Command('delete')
      .description('Delete an ontology and all its data')
      .showHelpAfterError()
      .argument('<name>', 'Ontology name')
      .option('-f, --force', 'Skip confirmation and force deletion')
      .action(async (name, options) => {
        try {
          if (!options.force) {
            console.log('\n' + separator());
            console.log(colors.status.warning('⚠️  WARNING: This action cannot be undone!'));
            console.log(separator());
            console.log(`\nThis will delete all data for ontology ${colors.concept.label(name)}`);
            console.log('\nUse ' + colors.ui.key('--force') + ' flag to confirm deletion\n');
            return;
          }

          const client = createClientFromEnv();
          const result = await client.deleteOntology(name, true);

          console.log('\n' + separator());
          console.log(colors.status.success(`✓ Deleted ontology "${result.ontology}"`));
          console.log(separator());
          console.log(`  ${colors.ui.key('Sources deleted:')} ${coloredCount(result.sources_deleted)}`);
          if (result.orphaned_concepts_deleted > 0) {
            console.log(`  ${colors.ui.key('Orphaned concepts cleaned:')} ${coloredCount(result.orphaned_concepts_deleted)}`);
          }
          console.log('\n' + separator());
        } catch (error: any) {
          console.error(colors.status.error('✗ Failed to delete ontology'));
          console.error(colors.status.error(error.response?.data?.detail || error.message));
          process.exit(1);
        }
      })
  );

File: api/api/routes/ontology.py
"""
Ontology management endpoints for organizing knowledge domains.

Provides REST API access to:
- List all ontologies
- Get ontology details and statistics
- List files within an ontology
- Delete ontologies with orphan cleanup
"""

from fastapi import APIRouter, Depends, HTTPException, Query as QueryParam
import logging

from ..dependencies.auth import CurrentUser, require_role
from ..models.ontology import (
    OntologyListResponse,
    OntologyItem,
    OntologyInfoResponse,
    OntologyFilesResponse,
    OntologyFileInfo,
    OntologyDeleteRequest,
    OntologyDeleteResponse,
    OntologyRenameRequest,
    OntologyRenameResponse
)
from api.api.lib.age_client import AGEClient

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/ontology", tags=["ontology"])


def get_age_client() -> AGEClient:
    """Get AGE client instance"""
    return AGEClient()


@router.get("/", response_model=OntologyListResponse)
async def list_ontologies(
    current_user: CurrentUser
):
    """
    List all ontologies in the knowledge graph (ADR-060).

    Returns summary statistics for each ontology including
    file count, chunk count, and concept count.

    **Authentication:** Requires valid OAuth token

    Returns:
        OntologyListResponse with all ontologies

    Example:
        GET /ontology/
    """
    client = get_age_client()
    try:
        result = client._execute_cypher("""
            MATCH (s:Source)
            WITH DISTINCT s.document as ontology
            MATCH (src:Source {document: ontology})
            WITH ontology,
                 count(DISTINCT src) as source_count,
                 count(DISTINCT src.file_path) as file_count
            OPTIONAL MATCH (c:Concept)-[:APPEARS]->(s:Source {document: ontology})
            WITH ontology, source_count, file_count, count(DISTINCT c) as concept_count
            RETURN ontology, source_count, file_count, concept_count
            ORDER BY ontology
        """)

        ontologies = [
            OntologyItem(
                ontology=record['ontology'],
                source_count=record['source_count'],
                file_count=record['file_count'],
                concept_count=record['concept_count']
            )
            for record in (result or [])
        ]

        return OntologyListResponse(
            count=len(ontologies),
            ontologies=ontologies
        )

    except Exception as e:
        logger.error(f"Failed to list ontologies: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to list ontologies: {str(e)}")
    finally:
        client.close()


@router.get("/{ontology_name}", response_model=OntologyInfoResponse)
async def get_ontology_info(
    ontology_name: str,
    current_user: CurrentUser
):
    """
    Get detailed information about a specific ontology (ADR-060).

    Includes:
    - File count and list of files
    - Chunk (source) count
    - Concept count
    - Evidence instance count
    - Relationship count

    **Authentication:** Requires valid OAuth token

    Args:
        ontology_name: Name of the ontology

    Returns:
        OntologyInfoResponse with detailed statistics

    Raises:
        404: If ontology not found

    Example:
        GET /ontology/Research%20Papers
    """
    client = get_age_client()
    try:
        # Check if ontology exists
        exists_check = client._execute_cypher(
            f"MATCH (s:Source {{document: '{ontology_name}'}}) RETURN count(s) > 0 as ontology_exists",
            fetch_one=True
        )

        if not exists_check or not exists_check['ontology_exists']:
            raise HTTPException(status_code=404, detail=f"Ontology '{ontology_name}' not found")

        # Get statistics
        stats = client._execute_cypher(f"""
            MATCH (s:Source {{document: '{ontology_name}'}})
            WITH count(DISTINCT s) as source_count,
                 count(DISTINCT s.file_path) as file_count,
                 collect(DISTINCT s.file_path) as files
            OPTIONAL MATCH (c:Concept)-[:APPEARS]->(src:Source {{document: '{ontology_name}'}})
            WITH source_count, file_count, files, count(DISTINCT c) as concept_count
            OPTIONAL MATCH (i:Instance)-[:FROM_SOURCE]->(src:Source {{document: '{ontology_name}'}})
            WITH source_count, file_count, files, concept_count, count(DISTINCT i) as instance_count
            OPTIONAL MATCH (ontology_concept:Concept)-[:APPEARS]->(:Source {{document: '{ontology_name}'}})
            OPTIONAL MATCH (ontology_concept)-[r]->(other:Concept)
            RETURN source_count, file_count, files, concept_count, instance_count, count(r) as relationship_count
        """, fetch_one=True)

        statistics = {
            "source_count": stats['source_count'],
            "file_count": stats['file_count'],
            "concept_count": stats['concept_count'],
            "instance_count": stats['instance_count'],
            "relationship_count": stats['relationship_count']
        }

        # Filter out None values from files list
        files = [f for f in stats['files'] if f is not None]

        return OntologyInfoResponse(
            ontology=ontology_name,
            statistics=statistics,
            files=files
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get ontology info: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to get ontology info: {str(e)}")
    finally:
        client.close()


@router.get("/{ontology_name}/files", response_model=OntologyFilesResponse)
async def get_ontology_files(
    ontology_name: str,
    current_user: CurrentUser
):
    """
    List all files in a specific ontology with their statistics (ADR-060).

    **Authentication:** Requires valid OAuth token

    Args:
        ontology_name: Name of the ontology

    Returns:
        OntologyFilesResponse with file details

    Raises:
        404: If ontology not found or has no files

    Example:
        GET /ontology/Research%20Papers/files
    """
    client = get_age_client()
    try:
        result = client._execute_cypher(f"""
            MATCH (s:Source {{document: '{ontology_name}'}})
            WITH DISTINCT s.file_path as file_path
            WHERE file_path IS NOT NULL
            MATCH (src:Source {{document: '{ontology_name}', file_path: file_path}})
            WITH file_path, count(src) as chunk_count
            OPTIONAL MATCH (c:Concept)-[:APPEARS]->(s:Source {{document: '{ontology_name}', file_path: file_path}})
            WITH file_path, chunk_count, count(DISTINCT c) as concept_count
            RETURN file_path, chunk_count, concept_count
            ORDER BY file_path
        """)

        files = [
            OntologyFileInfo(
                file_path=record['file_path'],
                chunk_count=record['chunk_count'],
                concept_count=record['concept_count']
            )
            for record in (result or [])
        ]

        if not files:
            raise HTTPException(
                status_code=404,
                detail=f"No files found in ontology '{ontology_name}'"
            )

        return OntologyFilesResponse(
            ontology=ontology_name,
            count=len(files),
            files=files
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get ontology files: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to get ontology files: {str(e)}")
    finally:
        client.close()


@router.delete("/{ontology_name}", response_model=OntologyDeleteResponse)
async def delete_ontology(
    ontology_name: str,
    current_user: CurrentUser,
    _: None = Depends(require_role("admin")),
    force: bool = QueryParam(False, description="Skip confirmation and force deletion")
):
    """
    Delete an ontology and all its data (Admin only - ADR-060).

    **WARNING: This action cannot be undone!**

    Deletes:
    - All Source nodes belonging to the ontology
    - All Instance nodes linked to those sources
    - All DocumentMeta nodes for this ontology (ADR-051: provenance metadata)
    - Orphaned Concept nodes (concepts with no remaining sources)
    - All source embeddings for deleted sources (kg_api.source_embeddings)
    - All job records for this ontology (enables clean re-ingestion)
    - All Garage storage objects (images) for this ontology

    **Authentication:** Requires admin role

    Args:
        ontology_name: Name of the ontology to delete
        force: If True, skip existence check and proceed with deletion

    Returns:
        OntologyDeleteResponse with deletion statistics

    Raises:
        404: If ontology not found (when force=False)

    Example:
        DELETE /ontology/Test%20Ontology?force=true
    """
    from ..services.job_queue import get_job_queue

    client = get_age_client()
    queue = get_job_queue()
    try:
        # Check if ontology exists
        if not force:
            check = client._execute_cypher(f"""
                MATCH (s:Source {{document: '{ontology_name}'}})
                WITH count(s) as source_count
                OPTIONAL MATCH (c:Concept)-[:APPEARS]->(s:Source {{document: '{ontology_name}'}})
                RETURN source_count, count(DISTINCT c) as concept_count
            """, fetch_one=True)

            if not check or check['source_count'] == 0:
                raise HTTPException(
                    status_code=404,
                    detail=f"Ontology '{ontology_name}' not found"
                )

        # ADR-057: Clean up Garage objects before deleting sources
        # Query for all Garage object keys in this ontology
        storage_objects_result = client._execute_cypher(f"""
            MATCH (s:Source {{document: '{ontology_name}'}})
            WHERE s.storage_key IS NOT NULL
            RETURN s.storage_key as storage_key
        """)

        storage_keys_to_delete = []
        if storage_objects_result:
            for row in storage_objects_result:
                if row.get('storage_key'):
                    storage_keys_to_delete.append(row['storage_key'])

        # Delete Garage objects
        if storage_keys_to_delete:
            try:
                from ..lib.garage_client import get_garage_client
                garage_client = get_garage_client()

                deleted_count = 0
                for object_key in storage_keys_to_delete:
                    try:
                        garage_client.delete_image(object_key)
                        deleted_count += 1
                    except Exception as e:
                        logger.warning(f"Failed to delete Garage object {object_key}: {e}")

                if deleted_count > 0:
                    logger.info(f"Deleted {deleted_count} images from Garage for ontology '{ontology_name}'")
            except Exception as e:
                logger.warning(f"Failed to initialize Garage client for cleanup: {e}")

        # Delete instances linked to sources in this ontology
        client._execute_cypher(f"""
            MATCH (i:Instance)-[:FROM_SOURCE]->(s:Source {{document: '{ontology_name}'}})
            DETACH DELETE i
        """)

        # IMPORTANT: Capture source_ids BEFORE cascade deleting from graph.
        # After graph nodes are deleted, we lose the ability to know which source_ids
        # belonged to this ontology - they're just opaque hashes in the relational table.
        # This pattern: capture references → cascade delete → cleanup related tables.
        source_ids_result = client._execute_cypher(f"""
            MATCH (s:Source {{document: '{ontology_name}'}})
            RETURN s.source_id as source_id
        """)
        source_ids = [row['source_id'] for row in (source_ids_result or []) if row.get('source_id')]

        # Delete sources from graph
        result = client._execute_cypher(f"""
            MATCH (s:Source {{document: '{ontology_name}'}})
            DETACH DELETE s
            RETURN count(s) as deleted_count
        """, fetch_one=True)

        sources_deleted = result['deleted_count'] if result else 0

        # Delete source_embeddings for the deleted sources
        if source_ids:
            try:
                conn = client.conn
                with conn.cursor() as cur:
                    # Use ANY to match source_ids in the list
                    cur.execute("""
                        DELETE FROM kg_api.source_embeddings
                        WHERE source_id = ANY(%s)
                    """, (source_ids,))
                    embeddings_deleted = cur.rowcount
                    conn.commit()
                    if embeddings_deleted > 0:
                        logger.info(f"Deleted {embeddings_deleted} source embeddings for ontology '{ontology_name}'")
            except Exception as e:
                logger.warning(f"Failed to delete source embeddings: {e}")

        # ADR-051: Delete DocumentMeta nodes for this ontology
        # This ensures cascade deletion of provenance metadata
        doc_meta_result = client._execute_cypher(f"""
            MATCH (d:DocumentMeta {{ontology: '{ontology_name}'}})
            DETACH DELETE d
            RETURN count(d) as deleted_count
        """, fetch_one=True)

        doc_meta_deleted = doc_meta_result['deleted_count'] if doc_meta_result else 0
        if doc_meta_deleted > 0:
            logger.info(f"Deleted {doc_meta_deleted} DocumentMeta nodes for ontology '{ontology_name}'")

        # Clean up orphaned concepts (concepts with no sources)
        # AGE doesn't support WHERE NOT with patterns, use OPTIONAL MATCH instead
        orphaned_result = client._execute_cypher("""
            MATCH (c:Concept)
            OPTIONAL MATCH (c)-[:APPEARS]->(s:Source)
            WITH c, s
            WHERE s IS NULL
            DETACH DELETE c
            RETURN count(c) as orphaned_count
        """, fetch_one=True)

        orphaned_count = orphaned_result['orphaned_count'] if orphaned_result else 0

        # Delete job records for this ontology to allow clean re-ingestion
        jobs_deleted = queue.delete_jobs_by_ontology(ontology_name)
        if jobs_deleted > 0:
            logger.info(f"Deleted {jobs_deleted} job records for ontology '{ontology_name}'")

        return OntologyDeleteResponse(
            ontology=ontology_name,
            deleted=True,
            sources_deleted=sources_deleted,
            orphaned_concepts_deleted=orphaned_count
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to delete ontology: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to delete ontology: {str(e)}")
    finally:
        client.close()


@router.post("/{ontology_name}/rename", response_model=OntologyRenameResponse)
async def rename_ontology(
    ontology_name: str,
    request: OntologyRenameRequest,
    current_user: CurrentUser,
    _: None = Depends(require_role("admin"))
):
    """
    Rename an ontology (Admin only - ADR-060).

    Updates all Source nodes' document property from old_name to new_name.
    This operation is fast and safe - only affects Source nodes in the specified ontology.

    **Authentication:** Requires admin role

    Args:
        ontology_name: Current ontology name
        request: Rename request with new_name

    Returns:
        OntologyRenameResponse with operation statistics

    Raises:
        404: If old ontology not found
        409: If new ontology name already exists

    Example:
        POST /ontology/Old%20Name/rename
        {
          "new_name": "New Name"
        }
    """
    client = get_age_client()
    try:
        # Perform rename via AGE client
        try:
            result = client.rename_ontology(ontology_name, request.new_name)

            return OntologyRenameResponse(
                old_name=ontology_name,
                new_name=request.new_name,
                sources_updated=result["sources_updated"],
                success=True
            )

        except ValueError as ve:
            # ValueError is raised for existence checks
            error_msg = str(ve)
            if "does not exist" in error_msg:
                raise HTTPException(status_code=404, detail=error_msg)
            elif "already exists" in error_msg:
                raise HTTPException(status_code=409, detail=error_msg)
            else:
                raise HTTPException(status_code=400, detail=error_msg)

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to rename ontology: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to rename ontology: {str(e)}")
    finally:
        client.close()
